### PR TITLE
Latency metrics

### PR DESCRIPTION
**Pull request by tomkordic**
Latency metrics v1.0

### CODE REVIEW

File: MuxExoPlayer/src/main/java/com/mux/stats/sdk/muxstats/MuxBaseExoPlayer.java [hunk 1]
```diff
@@ -72,7 +75,7 @@
  * of {@link ExoPlayer} version and make sure that all events are passed to {@link MuxStats} in
  * the correct order and with appropriate statistics.
  */
-public class MuxBaseExoPlayer extends EventBus implements IPlayerListener {
+public abstract class MuxBaseExoPlayer extends EventBus implements IPlayerListener {
```
Comment (skidder) [inline MuxExoPlayer/src/main/java/com/mux/stats/sdk/muxstats/MuxBaseExoPlayer.java:78]: Making this class `abstract` feels like the right thing to do, given that it's always extended by Exoplayer-version-specific subclasses. 👍

File: MuxExoPlayer/src/main/java/com/mux/stats/sdk/muxstats/MuxBaseExoPlayer.java [hunk 2]
```diff
@@ -661,6 +666,79 @@ public int getPlayerViewHeight() {
...
+    return -1L;
+  }
+
+  /**
+   * This is the time of the furthest position in the manifest as extrapolated from the PDT tags in
+   * the stream. Only available for DASH and HLS.
+   *
+   * @return time.
+   */
+  @RequiresApi(api = VERSION_CODES.O)
```
Comment (skidder) [inline MuxExoPlayer/src/main/java/com/mux/stats/sdk/muxstats/MuxBaseExoPlayer.java]: What's the significance of setting this to `RequiresApi` with version zero? Is this necessary?

File: MuxExoPlayer/src/r2_11_1/java/com/mux/stats/sdk/muxstats/MuxStatsExoPlayer.java [hunk 3]
```diff
@@ -111,6 +112,48 @@ public MuxStatsExoPlayer(Context ctx, ExoPlayer player, String playerName,
...
 
+  /**
+   * Extracts the tag value from live HLS segment, returns -1 if it is not an HLS stream, not a live
+   * playback.
+   *
+   * @param tagName name of the tag to extract from the HLS manifest.
+   * @return tag value if tag is found and we are playing HLS live stream, -1 string otherwise.
+   */
+  @Override
+  protected String parseHlsManifestTag(String tagName) {
```
Comment (skidder) [inline MuxExoPlayer/src/r2_11_1/java/com/mux/stats/sdk/muxstats/MuxStatsExoPlayer.java:123]: Can you pull this function implementation up into the `MuxBaseExoPlayer` class and override with the no-op implementation for the player versions where it's not supported?

**Review by skidder — CHANGES_REQUESTED** [review]

File: MuxExoPlayer/src/main/java/com/mux/stats/sdk/muxstats/MuxBaseExoPlayer.java (hunk 2 above)
Comment (daytime-em) [reply MuxExoPlayer/src/main/java/com/mux/stats/sdk/muxstats/MuxBaseExoPlayer.java]: I think that's an upper-case O, API level 26.

Same question though: I can't see what min-level-26 APIs are being invoked here

File: MuxExoPlayer/src/main/java/com/mux/stats/sdk/muxstats/MuxBaseExoPlayer.java [hunk 4]
```diff
@@ -1095,6 +1190,11 @@ public long getElapsedRealtime() {
       return elapsedRealtime();
     }
 
+    @Override
+    public void outputLog(LogPriority logPriority, String s, String s1) {
+
```
Comment (daytime-em) [inline MuxExoPlayer/src/main/java/com/mux/stats/sdk/muxstats/MuxBaseExoPlayer.java]:
```suggestion
switch (logPriority) {
        case ERROR:
          Log.e(tag, msg);
          break;
        case WARN:
          Log.w(tag, msg);
          break;
        case INFO:
          Log.i(tag, msg);
          break;
        case DEBUG:
          Log.d(tag, msg);
          break;
        case VERBOSE:
        default: // fall-through
          Log.v(tag, msg);
          break;
      }
```

File: build.gradle [hunk 5]
```diff
@@ -33,7 +33,7 @@ buildscript {
 //
 // It must be false for fat aar generation of MuxExoPlayer to work
 // This is a bug/problem with the fat aar mechanism
-def inLocalDevModeFlag = false
+def inLocalDevModeFlag = true
```
Comment (daytime-em) [inline build.gradle]:
```suggestion
def inLocalDevModeFlag = false
```
... right?

**Review by daytime-em — CHANGES_REQUESTED** [review]

File: build.gradle (hunk 5 above)
Comment (nbirkenshaw-mux) [reply build.gradle]: I have to admit I'm slightly confused by the ongoing presence of inLocalDevModeFlag - it's not in the master branch.

Has something gone wrong with the merging etc. here or it's being reintroduced deliberately?

Comment (skidder) [reply build.gradle]: I imagine this was included by accident, setting it to `false` seems right

Comment (daytime-em) [reply build.gradle]: I can clean it up in the next version, along with applying the new approach for the module dependencies.

File: MuxExoPlayer/src/r2_11_1/java/com/mux/stats/sdk/muxstats/MuxStatsExoPlayer.java (hunk 3 above)
Comment (daytime-em) [reply MuxExoPlayer/src/r2_11_1/java/com/mux/stats/sdk/muxstats/MuxStatsExoPlayer.java:123]: I didn't do this one for time's sake. I'll make a Story to follow up.

**Pull request merged**